File: whatsapp_bot/simulate_persona.js
const fs = require("fs");
const path = require("path");
const axios = require("axios");
const dns = require("dns");

dns.setDefaultResultOrder("ipv4first");

const OLLAMA_URL = process.env.OLLAMA_URL || "http://127.0.0.1:11434/api/generate";
const RAG_URL = process.env.RAG_URL || "http://127.0.0.1:5050/query";
const OLLAMA_MODEL = process.env.OLLAMA_MODEL_NAME || "target_persona_clone";
const TARGET_PERSONA = process.env.TARGET_PERSONA_NAME || "Target Persona";
const CHAT_PARTNER = process.env.CHAT_PARTNER_NAME || "Chat Partner";
const CHAT_LOG_PATH = process.env.WHATSAPP_EXPORT_PATH
  ? path.resolve(process.env.WHATSAPP_EXPORT_PATH)
  : path.join(__dirname, "..", "whatsapp_chat.txt");

const getRagContext = async (query) => {
  try {
    const res = await axios.post(RAG_URL, { query, top_k: 3 }, { timeout: 5000 });
    const results = res.data.results || res.data || [];
    return Array.isArray(results) ? results.join("\n") : "";
  } catch {
    return "";
  }
};

const buildPrompt = (userQuery, ragHint) => {
  let systemMsg = `# MODULE 1: IDENTITY
You are ${TARGET_PERSONA}. You are chatting over WhatsApp with ${CHAT_PARTNER}.
Match the target persona's style from the fine-tuning data.
Keep replies concise, natural, and not assistant-like.
Do not invent people, stories, memories, or events.`;

  if (ragHint) {
    systemMsg += `\n# STYLISTIC MEMORIES\nUse only if relevant:\n${ragHint}`;
  }

  return `<|im_start|>system\n${systemMsg}<|im_end|>\n<|im_start|>user\n${userQuery}<|im_end|>\n<|im_start|>assistant\n`;
};

const simulate = async () => {
  console.log("\nCOPYHERR Simulation Engine\n");

  if (!fs.existsSync(CHAT_LOG_PATH)) {
    console.error(`Chat log not found: ${CHAT_LOG_PATH}`);
    return;
  }

  const content = fs.readFileSync(CHAT_LOG_PATH, "utf8");
  const lines = content.split("\n");
  const partnerLines = lines
    .filter((line) => line.includes(`- ${CHAT_PARTNER}:`))
    .map((line) => line.split(`- ${CHAT_PARTNER}:`)[1].trim())
    .filter((line) => line.length > 10 && line.length < 100);

  const samples = [];
  for (let i = 0; i < Math.min(5, partnerLines.length); i++) {
    samples.push(partnerLines[Math.floor(Math.random() * partnerLines.length)]);
  }
  samples.push("hi");
  samples.push("how are you?");

  for (const sample of samples) {
    console.log(`\nTest Query > ${sample}`);
    const ragHint = await getRagContext(sample);
    const prompt = buildPrompt(sample, ragHint);

    try {
      const response = await axios.post(OLLAMA_URL, {
        model: OLLAMA_MODEL,
        prompt,
        stream: false,
        options: { temperature: 0.3 },
      });

      const fullText = response.data.response || "";
      const thoughtMatch = fullText.match(/<thought>([\s\S]*?)<\/thought>/);
      const thought = thoughtMatch ? thoughtMatch[1] : "No thinking logged";
      const reply = fullText.replace(/<thought>[\s\S]*?<\/thought>/gi, "").trim();

      console.log(`Thought: ${thought}`);
      console.log(`${TARGET_PERSONA}: ${reply}`);
    } catch (err) {
      console.error(`Model error: ${err.message}`);
    }
  }
};

simulate();

File: whatsapp_bot/simulate_persona.py
import os
import random
import re

import requests

OLLAMA_URL = os.environ.get("OLLAMA_URL", "http://localhost:11434/api/generate")
RAG_URL = os.environ.get("RAG_URL", "http://127.0.0.1:5050/query")
OLLAMA_MODEL = os.environ.get("OLLAMA_MODEL_NAME", "target_persona_clone")
TARGET_PERSONA = os.environ.get("TARGET_PERSONA_NAME", "Target Persona")
CHAT_PARTNER = os.environ.get("CHAT_PARTNER_NAME", "Chat Partner")

PWD = os.path.dirname(os.path.abspath(__file__))
export_path = os.environ.get("WHATSAPP_EXPORT_PATH", "whatsapp_chat.txt")
CHAT_LOG_PATH = export_path if os.path.isabs(export_path) else os.path.join(PWD, "..", export_path)


def get_rag_context(query):
    try:
        response = requests.post(RAG_URL, json={"query": query, "top_k": 3}, timeout=5)
        data = response.json()
        results = data.get("results", data if isinstance(data, list) else [])
        return "\n".join(results)
    except Exception as exc:
        print(f"[RAG Error] {exc}")
        return ""


def build_prompt(user_query, rag_hint):
    system_msg = f"""# MODULE 1: IDENTITY
You are {TARGET_PERSONA}. You are chatting over WhatsApp with {CHAT_PARTNER}.
Match the target persona's style from the fine-tuning data.
Keep replies concise, natural, and not assistant-like.
Do not invent people, stories, memories, or events."""

    if rag_hint:
        system_msg += f"\n# STYLISTIC MEMORIES\nUse only if relevant:\n{rag_hint}"

    return (
        f"<|im_start|>system\n{system_msg}<|im_end|>\n"
        f"<|im_start|>user\n{user_query}<|im_end|>\n"
        f"<|im_start|>assistant\n"
    )


def simulate():
    print("\nCOPYHERR Simulation Engine\n")

    if not os.path.exists(CHAT_LOG_PATH):
        print(f"[Error] Chat log not found at {CHAT_LOG_PATH}")
        return

    with open(CHAT_LOG_PATH, "r", encoding="utf-8") as file:
        lines = file.readlines()

    partner_lines = [
        line.split(f"- {CHAT_PARTNER}:")[1].strip()
        for line in lines
        if f"- {CHAT_PARTNER}:" in line
    ]
    partner_lines = [line for line in partner_lines if 10 < len(line) < 100]

    samples = random.sample(partner_lines, min(5, len(partner_lines))) if partner_lines else []
    samples.extend(["hi", "how are you?"])

    for sample in samples:
        print(f"\nTest Query > {sample}")
        rag_hint = get_rag_context(sample)
        prompt = build_prompt(sample, rag_hint)

        try:
            response = requests.post(
                OLLAMA_URL,
                json={
                    "model": OLLAMA_MODEL,
                    "prompt": prompt,
                    "stream": False,
                    "options": {"temperature": 0.5},
                },
                timeout=30,
            )

            full_text = response.json()["response"]
            thought_match = re.search(r"<thought>(.*?)</thought>", full_text, re.DOTALL)
            thought = thought_match.group(1).strip() if thought_match else "No thinking logged"
            reply = re.sub(r"<thought>.*?</thought>", "", full_text, flags=re.DOTALL).strip()

            print(f"Thought: {thought}")
            print(f"{TARGET_PERSONA}: {reply}")
        except Exception as exc:
            print(f"[Model Error] {exc}")


if __name__ == "__main__":
    simulate()
